osu.Game/Screens/Play/Player.cs

```diff
@@ -1115,8 +1115,7 @@ public override bool OnExiting(ScreenExitEvent e)
                 if (!GameplayState.HasPassed && !GameplayState.HasFailed)
                     GameplayState.HasQuit = true;
 
-                // if arriving here and the results screen preparation task hasn't run, it's safe to say the user has not completed the beatmap.
-                if (prepareScoreForDisplayTask == null && DrawableRuleset.ReplayScore == null)
+                if (!GameplayState.HasPassed && DrawableRuleset.ReplayScore == null)
```

Comment:
Seems correct, but I'd propose a slight simplification:

```diff
diff --git a/osu.Game/Screens/Play/Player.cs b/osu.Game/Screens/Play/Player.cs
index 18d0a65d7a..a1ec0b3167 100644
--- a/osu.Game/Screens/Play/Player.cs
+++ b/osu.Game/Screens/Play/Player.cs
@@ -1110,12 +1110,12 @@ public override bool OnExiting(ScreenExitEvent e)
             failAnimationContainer?.Stop();
             PauseOverlay?.StopAllSamples();
 
-            if (LoadedBeatmapSuccessfully)
+            if (LoadedBeatmapSuccessfully && !GameplayState.HasPassed)
             {
-                if (!GameplayState.HasPassed && !GameplayState.HasFailed)
+                if (!GameplayState.HasFailed)
                     GameplayState.HasQuit = true;
 
-                if (!GameplayState.HasPassed && DrawableRuleset.ReplayScore == null)
+                if (DrawableRuleset.ReplayScore == null)
                     ScoreProcessor.FailScore(Score.ScoreInfo);
             }
 

```

Reply:
Sure, applied.

I also threw in some extra assertions in https://github.com/ppy/osu/pull/25264/commits/dc7f5cd6edc94275a56b3b5cdb5766c76c84181c for added peace of mind. Not sure if that's what you were imagining when this was brought up in stream chat.

Reply:
Turns out I went a bit overboard with the assertion; fixed in https://github.com/ppy/osu/pull/25264/commits/35f30d6135b29b73b68d861bb68564cc951f1a27.